src/lib389/lib389/cli_idm/account.py

```diff
@@ -118,6 +119,52 @@ def subtree_status(inst, basedn, log, args):
         _print_entry_status(status, entry.dn, log)
 
 
+def bulk_update(inst, basedn, log, args):
+    basedn = _get_dn_arg(args.basedn, msg="Enter basedn to search")
+    search_filter = "(objectclass=*)"
+    scope = ldap.SCOPE_SUBTREE
+    if args.scope == "one":
+        scope = ldap.SCOPE_ONELEVEL
+    if args.filter:
+        search_filter = args.filter
+
+    entry_list = Accounts(inst, basedn).filter(search_filter, scope=scope)
+    if not entry_list:
+        raise ValueError(f"No entries were found.")
+
+    failed_list = []
+    success_list = []
+    for entry in entry_list:
+        if entry.dn.lower() == basedn.lower():
+            # skip parent
+            continue
+        try:
+            _generic_modify_dn(inst, basedn, log.getChild('_generic_modify_dn'), MANY, entry.dn, args)
+            success_list.append(entry.dn)
+        except ldap.LDAPError as e:
+            if "desc" in e.args[0]:
+                failed_list.append(entry.dn + f" ({e.args[0]['desc']})")
+            else:
+                failed_list.append(entry.dn + f" ({str(e)})")
+            if args.stop:
+                raise ValueError(f"Failed to update entry ({entry.dn}), error: {str(e)}")
+
+    log.info(f"Updated {len(success_list)} entries")
+    if len(failed_list) > 0:
+        log.info(f"Failed to update {len(failed_list)} entries")
```

Comment:
I get an odd "log" issue when it says 'failed' and 'updated' at the same time. But it's actually updated correctly so `Failed to update 5 entries` is false, actually.

Steps to reproduce:
1. `DEBUGGING=yes py.test-3 -s -v dirsrvtests/tests/suites/clu/dsidm_bulk_update_test.py`
2. `dsctl standalone1 start`
3. `dsidm -v standalone1 account bulk_update dc=example,dc=com replace:cn:new_cn`

```
Successfully modified uid=demo_user,ou=people,dc=example,dc=com
Successfully modified uid=testuser0,ou=People,dc=example,dc=com
Successfully modified uid=testuser1,ou=People,dc=example,dc=com
Successfully modified uid=testuser2,ou=People,dc=example,dc=com
Successfully modified uid=testuser3,ou=People,dc=example,dc=com
Successfully modified uid=testuser4,ou=People,dc=example,dc=com
Updated 6 entries
Failed to update 5 entries
```

4. `dsidm standalone1 -b dc=example,dc=com account get-by-dn`

```
dn: uid=testuser4,ou=People,dc=example,dc=com
cn: new_cn
gidNumber: 1004
homeDirectory: /home/testuser4
objectClass: top
objectClass: account
objectClass: posixaccount
objectClass: inetOrgPerson
objectClass: organizationalPerson
objectClass: nsMemberOf
objectClass: nsAccount
objectClass: person
sn: user4
uid: testuser4
uidNumber: 1004
```

Reply:
Actually this is correct.  If you use the "--details" option you can see what entries were not updated:

```
Success [1]: uid=demo_user,ou=people,dc=example,dc=com
Success [2]: uid=testuser0,ou=People,dc=example,dc=com
Success [3]: uid=testuser1,ou=People,dc=example,dc=com
Success [4]: uid=testuser2,ou=People,dc=example,dc=com
Success [5]: uid=testuser3,ou=People,dc=example,dc=com
Success [6]: uid=testuser4,ou=People,dc=example,dc=com
Failed [1]: ou=groups,dc=example,dc=com (Object class violation)
Failed [2]: ou=people,dc=example,dc=com (Object class violation)
Failed [3]: ou=permissions,dc=example,dc=com (Object class violation)
Failed [4]: ou=services,dc=example,dc=com (Object class violation)
Failed [5]: cn=demo_group,ou=Groups,dc=example,dc=com (Operation not allowed on RDN)
```

If you use a filter you can avoid this:

```
# dsidm  standalone1 account bulk_update dc=example,dc=com replace:cn:new_cn --details --filter "(uid=*)"
Successfully modified uid=demo_user,ou=people,dc=example,dc=com
Successfully modified uid=testuser0,ou=People,dc=example,dc=com
Successfully modified uid=testuser1,ou=People,dc=example,dc=com
Successfully modified uid=testuser2,ou=People,dc=example,dc=com
Successfully modified uid=testuser3,ou=People,dc=example,dc=com
Successfully modified uid=testuser4,ou=People,dc=example,dc=com
Updated 6 entries
-------------- Details --------------
Success [1]: uid=demo_user,ou=people,dc=example,dc=com
Success [2]: uid=testuser0,ou=People,dc=example,dc=com
Success [3]: uid=testuser1,ou=People,dc=example,dc=com
Success [4]: uid=testuser2,ou=People,dc=example,dc=com
Success [5]: uid=testuser3,ou=People,dc=example,dc=com
Success [6]: uid=testuser4,ou=People,dc=example,dc=com
```

Reply:
Okay, got it.
I have a couple of additional thoughts then :)

1. I think it is worth mentioning on the first line something like: `Using (objectClass=*) filter, etc... found 11 entries.
2. To report errors like `Failed [1]: ou=groups,dc=example,dc=com (Object class violation)` to the debug log level too. Because I've used `dsidm -v` option without `--details`, and it gave me information only about successes.

What do you think?

Reply:
> Okay, got it. I have a couple of additional thoughts then :)
> 
> 1. I think it is worth mentioning on the first line something like: `Using (objectClass=*) filter, etc... found 11 entries`.
> 2. Report errors like `Failed [1]: ou=groups,dc=example,dc=com (Object class violation)` to the debug log level too. Because I've used `dsidm -v` option without `--details`, and it gave me information only about successes.
> 
> What do you think?

Reply:
@droideck ok reworked things a bit.  I removed the "--details" option.  Now we always display what was updated and what wasn't:

```
# dsidm  standalone1 account bulk_update dc=example,dc=com replace:cn:new_cn  
Searching 'dc=example,dc=com' filter '(objectclass=*)' scope '2' ...
Found 12 matching entries.
Successfully modified uid=demo_user,ou=people,dc=example,dc=com
Successfully modified uid=testuser0,ou=People,dc=example,dc=com
Successfully modified uid=testuser1,ou=People,dc=example,dc=com
Successfully modified uid=testuser2,ou=People,dc=example,dc=com
Successfully modified uid=testuser3,ou=People,dc=example,dc=com
Successfully modified uid=testuser4,ou=People,dc=example,dc=com
Updates Finished.
Successfully updated 6 entries.
Failed to update 6 entries:
[1] dc=example,dc=com (Base DN Entry Skipped)
[2] ou=groups,dc=example,dc=com (Object class violation)
[3] ou=people,dc=example,dc=com (Object class violation)
[4] ou=permissions,dc=example,dc=com (Object class violation)
[5] ou=services,dc=example,dc=com (Object class violation)
[6] cn=demo_group,ou=Groups,dc=example,dc=com (Operation not allowed on RDN)
```

Reply:
Looks great!
Minor nitpick - should we change `scope '2'` to something like `scope 'sub'`? So it's aligned with the CLI parameters?

And I'll put another suggestion with another comment below.